test/test_autograd.py

```diff
@@ -5721,6 +5775,46 @@ def test_copy_(self, device):
             z = x.to(torch.bfloat16)
             self.assertTrue(z.requires_grad)
 
+    @onlyCUDA
+    def test_simple_reentrant_cross_device(self, device):
+        class ReentrantFunc(Function):
+            _cpu_mode = True
+            @staticmethod
+            def forward(ctx, x):
+                with torch.enable_grad():
+                    ctx.output_var = x * (x + 2) 
+                return ctx.output_var.detach()
+
+            @staticmethod
+            def backward(ctx, grad_output):
+                with torch.enable_grad():
+                    if ReentrantFunc._cpu_mode:
+                        new_param = torch.randn(2, 2, requires_grad=True)
+                        (new_param ** 2).sum().backward()
+                    else:
+                        new_param = torch.randn(2, 2, device=device, requires_grad=True)
+                        (new_param ** 2).sum().backward()
+                return grad_output
+
+        # Reentrant starts on GPU thread, finishs on GPU thread
+        x = torch.randn(2, 2, device=device, requires_grad=True)
+        out = ReentrantFunc.apply(x)
+        out.sum().backward()
+
+        # Reentrant starts on CPU thread, finishs on GPU thread
```

Comment:
Doesn't the backward technically finish on the CPU thread since the last thing we execute is AccumulateGrad for `x`?

Reply:
Yeah the root backward finishes on CPU as we execute `AccumulateGrad` for `x`, but I think here we want to test the `reentrant` backward calls starts on CPU and finish on GPU? For reentrant backward, it starts on CPU as the root backward on CPU, it then finishes it's backward on GPU since we run `AccumulateGrad` for `new_param` on GPU?